### PR TITLE
chore: convert compareArray() to assert.compareArray()

### PR DESCRIPTION
**Pull request by rwaldron**

### CODE REVIEW

File: harness/compareArray.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ compareArray.isSameValue = function(a, b) {
 };
 
 compareArray.format = function(spreadable) {
-  return `[${[...spreadable].map(String).join(', ')}]`;
+  return `[${[].map.call(spreadable, String).join(', ')}]`;
```
Comment (jugglinmike) [inline harness/compareArray.js:30]: Dunno if this will work--a "spreadable" is an iterable, right?

**Review by leobalter — DISMISSED** [review]

File: harness/compareArray.js (hunk 1 above)
Comment (leobalter) [reply harness/compareArray.js:30]: there is a whole problematic over this compareArray file and this format is being only used for the assertion message.

The assertion requires the values a and b (actual and expected) to be accessed in a for loop during the assertion, and the values would be accessed again for this formatting purposes. 

I believe this `[].map.call` is a conservative approach and I'd stick with it for discussing compareArray in a different PR.

File: harness/compareArray.js [hunk 2]
```diff
@@ -26,8 +26,8 @@ compareArray.isSameValue = function(a, b) {
   return a === b;
 };
 
-compareArray.format = function(array) {
-  return `[${array.map(String).join(', ')}]`;
+compareArray.format = function(thingWithALengthPropertyAndNumericIndices) {
+  return `[${[].map.call(thingWithALengthPropertyAndNumericIndices, String).join(', ')}]`;
```
Comment (rwaldron) [inline harness/compareArray.js:30]: @jugglinmike @leobalter how about now?

Comment (jugglinmike) [reply harness/compareArray.js:30]: Anything that doesn't suggest we expect `map` to exhaust an iterable.

Comment (leobalter) [reply harness/compareArray.js:30]: The original code was using map directly on the given array-like object. Now it only uses map on top of the same object. I'd rather advance any further modifications/improvements to this file into another PR, WDYT?

Comment (jugglinmike) [reply harness/compareArray.js:30]: Sure (I'm not requesting any further modifications)

File: harness/compareArray.js [hunk 3]
```diff
@@ -35,8 +35,11 @@ assert.compareArray = function(actual, expected, message) {
...
-    compareArray(actual, expected),
-    `Expected ${format(actual)} and ${format(expected)} to have the same contents. ${message}`
-  );
+  var result = compareArray(actual, expected);
+
+  // The following prevents actual and expected from being iterated and evaluated
+  // more than once unless absolutely necessary.
+  if (!result) {
+    assert(false, `Expected ${format(actual)} and ${format(expected)} to have the same contents. ${message}`);
+  }
```
Comment (leobalter) [inline harness/compareArray.js:44]: @jugglinmike Rick and I discussed about this assertion message that is now deferred to execute only if compareArray fails. That might _avoid_ conflicts for existing tests migrating from pure compareArray to assert.compareArray.

**Review by leobalter — APPROVED** [review]

**Comment by mhofman** [comment]
This PR mangled my previous PR #3213 (and possibly other changes!)

File: test/language/module-code/namespace/internals/define-own-property.js [hunk 4]
```diff
@@ -84,54 +84,33 @@ for (const key of ([...exported, Symbol.toStringTag])) {
...
-    "Object.freeze: " + String(i)
-  );
-}
-
-for (const key of exported) {
-  const desc = Object.getOwnPropertyDescriptor(ns, key);
-  assert.sameValue(desc.writable, true, String(key) + " writable");
-}
-
-assert(!Object.isFrozen(ns), "namespace object not frozen");
```
Comment (mhofman) [inline test/language/module-code/namespace/internals/define-own-property.js:137]: These lines were removed

**Pull request merged**